### DISCUSSION TITLE
Best way to derive equivalent of `fetcher.type` in React Router?

### DISCUSSION THREAD

**Discussion by edlaver** [remix-run/react-router#9390] [Q&A] [answered]
I've used `fetcher.type` extensively in my Remix code (mostly `type === "done"`), but saw that it's been removed from React Router: https://github.com/remix-run/react-router/commit/e4c9a2af58ddb7d83208ea5cade398d5d27af299

The commit message mentions that it was removed as it *"can be derived"*.

Is there any guidance on the best way to do this?

Do we need to track the previous `fetcher.state` and assume that `fetcher.type === "done"` is the equivalent of `fetcher.state` transitioning from `submitting|loading` => `idle`?

**Comment by kiliman** [comment, marked as answer]
This should help

|type|state|data|method|
|----|-----|----|------|
|init|idle|null|
|done|idle|not null|
|actionSubmission|submitting|null|POST|
|loaderSubmission|submitting|null|GET|
|actionReload|loading|not null|POST|
|actionRedirect _???not sure_|loading|not null|POST|
|normalReload|loading|not null|GET|

**Reply by edlaver** [reply]
Thanks @kiliman , that'll be easier to just check `state` and `data` than keeping track of a previous `state` value.

**Reply by rafael-lua** [reply]
Far from the future note, remix utils has a tool for this: https://github.com/sergiodxa/remix-utils?tab=readme-ov-file#derive-fetcher-type
